tests/unit/_experimental/dataset/test_dataset.py

```diff
@@ -52,6 +57,89 @@ def test__infer_datatype_value(uri: str, expected: str) -> None:
 
 
+def test__add_datatype() -> None:
+    df = pd.DataFrame(
+        dict(
+            locator=["s3://test.pdf", "https://test.png", "/home/test.mp4", "/tmp/test.pcd"],
+        ),
+    )
+    _add_datatype(df)
+    assert df[DATA_TYPE_FIELD].equals(
+        pd.Series([DatapointType.DOCUMENT, DatapointType.IMAGE, DatapointType.VIDEO, DatapointType.POINT_CLOUD]),
+    )
+
+
+def test__add_datatype__composite() -> None:
+    composite_dataset = pd.DataFrame(
+        {
+            "a.text": a_text,
+        },
+    )
+    _add_datatype(composite_dataset)
+
+    assert "a.text" not in composite_dataset
+    assert (composite_dataset[DATA_TYPE_FIELD] == DatapointType.COMPOSITE).all()
+    assert (composite_dataset["a"] == [{"text": text, DATA_TYPE_FIELD: DatapointType.TEXT} for text in a_text]).all()
+
+    similarity = [5.0, 3.799999952316284, 3.799999952316284]
+    composite_dataset = pd.DataFrame(
+        {
+            "a.text": a_text,
+            "b.text": b_text,
+            "similarity": similarity,
+        },
+    )
+    _add_datatype(composite_dataset)
+
+    assert (composite_dataset[DATA_TYPE_FIELD] == DatapointType.COMPOSITE).all()
+    assert (composite_dataset["similarity"] == similarity).all()
+    assert (composite_dataset["a"] == [{"text": text, DATA_TYPE_FIELD: DatapointType.TEXT} for text in a_text]).all()
+    assert (composite_dataset["b"] == [{"text": text, DATA_TYPE_FIELD: DatapointType.TEXT} for text in b_text]).all()
+
+
+def test__flatten_composite():
+    composite_dataset = pd.DataFrame(
+        {
+            "a": [{"text": text, DATA_TYPE_FIELD: DatapointType.TEXT.value} for text in a_text],
+            "b": [{"text": text, DATA_TYPE_FIELD: DatapointType.TEXT.value} for text in b_text],
+            "c": [dict(text=a + b) for a, b in zip(a_text, b_text)],  # Should not flatten because no DATA_TYPE_FIELD
+            DATA_TYPE_FIELD: DatapointType.COMPOSITE,
+        },
+    )
+    composite_dataset = _flatten_composite(composite_dataset)
+
+    expected = pd.DataFrame(
+        {
+            "a.text": a_text,
+            f"a.{DATA_TYPE_FIELD}": DatapointType.TEXT,
+            "b.text": b_text,
+            f"b.{DATA_TYPE_FIELD}": DatapointType.TEXT,
+            "c": [dict(text=a + b) for a, b in zip(a_text, b_text)],
+            DATA_TYPE_FIELD: DatapointType.COMPOSITE,
+        },
+    )
+    assert (composite_dataset.sort_index(axis=1) == expected.sort_index(axis=1)).all().all()
+
+
+def test__add_datatype__invalid():
+    composite_dataset = pd.DataFrame(
+        {
+            "a": [i for i in range(5)],
+            "a.locator": [f"{i}.png" for i in range(5)],
+        },
+    )
+    with pytest.raises(InputValidationError):
+        _add_datatype(composite_dataset)
+
+    composite_dataset = pd.DataFrame(
+        {
+            "too.many.dots": [i for i in range(5)],
+        },
+    )
+    with pytest.raises(InputValidationError):
+        _add_datatype(composite_dataset)
+
+
```

Comment:
I added this test

```
def test__datapoint_dataframe__serde_composite() -> None:
    datapoints = [
        {
            "category": "outdoor" if i < 5 else "indoor",
            "is_same": True if i % 3 else False,
            "a.locator": f"https://test-iamge-{i}.png",
            "a.width": 500 + i,
            "a.height": 400 + i,
            "a.bboxes": [
                LabeledBoundingBox(label="car", top_left=[i, i], bottom_right=[i + 50, i + 50])
                for i in range(random.randint(2, 6))
            ],
            "a.label": ScoredClassificationLabel(label="dog", score=0.1 + i * 0.05),
            "b.locator": f"https://test-iamge-{i}.png",
            "b.width": 500 + i,
            "b.height": 400 + i,
            "b.bboxes": [
                LabeledBoundingBox(label="car", top_left=[i, i], bottom_right=[i + 50, i + 50])
                for i in range(random.randint(2, 6))
            ],
            "b.label": ScoredClassificationLabel(label="cat", score=0.1 + i * 0.05),
        }
        for i in range(10)
    ]
    df = pd.DataFrame(datapoints)
    df_expected = pd.DataFrame(
        dict(
            datapoint=[
                dict(
                    category=dp["category"],
                    is_same=dp["is_same"],
                    a={
                        DATA_TYPE_FIELD: DatapointType.IMAGE,
                        "label": dp["a.label"]._to_dict(),
                        "width": dp["a.width"],
                        "height": dp["a.height"],
                        "locator": dp["a.locator"],
                        "bboxes": [bbox._to_dict() for bbox in dp["a.bboxes"]],
                    },
                    b={
                        DATA_TYPE_FIELD: DatapointType.IMAGE,
                        "label": dp["b.label"]._to_dict(),
                        "width": dp["b.width"],
                        "height": dp["b.height"],
                        "locator": dp["b.locator"],
                        "bboxes": [bbox._to_dict() for bbox in dp["b.bboxes"]],
                    },
                    data_type=DatapointType.COMPOSITE,
                )
                for dp in datapoints
            ],
        ),
    )
    df_serialized = _to_serialized_dataframe(df, column=COL_DATAPOINT)

    assert df_serialized[COL_DATAPOINT].apply(json.loads).equals(df_expected[COL_DATAPOINT])

    df_expected = pd.DataFrame(
        dict(
            datapoint=[
                {
                    "category": dp["category"],
                    "is_same": dp["is_same"],
                    "a.label": ClassificationLabel(label=dp["a.label"].label, score=dp["a.label"].score),
                    "a.width": dp["a.width"],
                    "a.height": dp["a.height"],
                    "a.locator": dp["a.locator"],
                    "a.bboxes": [
                        BoundingBox(label=bbox.label, top_left=bbox.top_left, bottom_right=bbox.bottom_right)
                        for bbox in dp["a.bboxes"]],
                    "b.label": ClassificationLabel(label=dp["b.label"].label, score=dp["b.label"].score),
                    "b.width": dp["b.width"],
                    "b.height": dp["b.height"],
                    "b.locator": dp["b.locator"],
                    "b.bboxes": [
                        BoundingBox(label=bbox.label, top_left=bbox.top_left, bottom_right=bbox.bottom_right)
                        for bbox in dp["b.bboxes"]],
                }
                for dp in datapoints
            ],
        ),
    )
    df_deserialized = _to_deserialized_dataframe(df_serialized, column=COL_DATAPOINT)
    assert_frame_equal(df_deserialized, df_expected)
```

the final deserialization looks to have flattened `a.label` and `b.label`, which should remain `ClassificationLabel`.

Reply:
Thanks! This test uncovered a few issues. The recursive flattening is fixed by adding `max_level=0` option to https://pandas.pydata.org/docs/reference/api/pandas.json_normalize.html.

Another issue this uncovered is LabeledBoundingBox being cast to BoundingBox during deserialization. It appears many of the [subtypes (eg. LabeledBoundingBox, ScoredBoundingBox) are missing annotation types](https://github.com/kolenaIO/kolena/blob/trunk/kolena/workflow/annotation.py#L106) and being cast to their parent class due to inheriting the annotation label.
```
(Pdb) df_deser['a.bboxes'][0]
[BoundingBox(top_left=(0.0, 0.0), bottom_right=(50.0, 50.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0), BoundingBox(top_left=(1.0, 1.0), bottom_right=(51.0, 51.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0), BoundingBox(top_left=(2.0, 2.0), bottom_right=(52.0, 52.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0)]
(Pdb) df_orig['a.bboxes'][0]
[LabeledBoundingBox(top_left=(0.0, 0.0), bottom_right=(50.0, 50.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car'), LabeledBoundingBox(top_left=(1.0, 1.0), bottom_right=(51.0, 51.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car'), LabeledBoundingBox(top_left=(2.0, 2.0), bottom_right=(52.0, 52.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car')]
```

Reply:
For the 2nd point, it is currently a known limitation for the tradeoff for ease of use, as we don't have enough information to deserialize to subtypes.